library/core/tests/alloc.rs

```diff
@@ -12,6 +13,49 @@ fn const_unchecked_layout() {
     assert_eq!(Some(DANGLING), NonNull::new(ptr::invalid_mut(ALIGN)));
 }
 
+#[test]
+fn layout_round_up_to_align_edge_cases() {
+    const MAX_SIZE: usize = isize::MAX as usize;
+
+    for shift in 0..usize::BITS {
+        let align = 1_usize << shift;
+        let edge = (MAX_SIZE + 1) - align;
+        let low = edge.saturating_sub(10);
+        let high = edge.saturating_add(10);
+        assert!(Layout::from_size_align(low, align).is_ok());
+        assert!(Layout::from_size_align(high, align).is_err());
+        for size in low..=high {
+            assert_eq!(
+                Layout::from_size_align(size, align).is_ok(),
+                size.next_multiple_of(align) <= MAX_SIZE,
+            );
+        }
+    }
+}
+
+#[test]
+fn layout_array_edge_cases() {
+    for_type::<i64>();
+    for_type::<[i32; 0b10101]>();
+    for_type::<[u8; 0b1010101]>();
+
+    // Make sure ZSTs don't lead to divide-by-zero
+    assert_eq!(Layout::array::<()>(usize::MAX).unwrap(), Layout::from_size_align(0, 1).unwrap(),);
```

Comment:
Nit: trailing comma snuck in here

Reply:
Ah, I'd written them as multi-line and rustfmt made them one-line but didn't remove the trailing comma.  Fixed.

Reply:
Yeah, unfortunately rustfmt is conservative here and doesn't know that the trailing `,` is meaningless. Without doing name lookup it's impossible for it to know for sure; heuristically assuming it's `std::assert_eq!` is fine for formatting, but insufficient to justify changing what tokens appear in the bang-macro.

(Theoretically the same could be said for attribute and derive macros, but those have a much stronger implication of not caring about trivial token adjustments like trailing commas. Also procedural bang macros could also use whitespace information by span info... it's a game of what can you get away with.)